Cereproc to daisy

carl-textalk opened the pull request:
Add Cereproc TTS adapter to daisy-pipeline.

carl-textalk commented on modules/tts/tts-adapters/tts-adapter-cereproc/src/main/java/org/daisy/pipeline/tts/cereproc/impl/CereProcEngine.java:
```diff
@@ -0,0 +1,205 @@
+
+	@Override
+	public Collection<Voice> getAvailableVoices() throws SynthesisException, InterruptedException {
+		// FIXME: don't hard code
```
I guess this needs to be fixed.

bertfrees commented:
@carl-textalk I cleaned up the PR and pushed to your branch. As agreed I squashed your commits. I did a couple small fixes but will ask you to do some other things.

carl-textalk commented:
> @carl-textalk I cleaned up the PR and pushed to your branch. As agreed I squashed your commits. I did a couple small fixes but will ask you to do some other things.

Great!

bertfrees reviewed (changes requested):
@carl-textalk I have found some issues. Can you please have a look?

By the way I'm also seeing a lot of formatting/style issues: bad indentation, unwanted or missing spaces, tabs, or newlines, imports not grouped or ordered correctly, things like that. But I think it's easiest if I go over the whole code and fix those formatting issues myself before I merge.

carl-textalk commented on modules/tts/tts-adapters/tts-adapter-cereproc/src/main/java/org/daisy/pipeline/tts/cereproc/impl/util/UCharReplacer.java:
```diff
@@ -0,0 +1,201 @@
+	 * @return a character sequence whose codepoints may or may not
+	 * have been replaced by substitution characters, depending on settings and substitution success.
+	 */
+	public CharSequence replace(CharSequence input) {
```
Can you elaborate, what is not thread safe in that function?

bertfrees replied:
I'm seeing at least one issue and that is that `mSubstitutionTable` and `mSubstitutionTablesIterator` are global variables. But there could be more issues, I don't know. You need to check. E.g. I don't know about the UCI classes.

AlexanderHugestrand replied:
Carl, just add the word "synchronized" to all methods that access the members of the class, and the entire class will be thread safe. This method in particular doesn't need that though, since it doesn't access any member directly. It calls the method getSubstitutionChars(), which in turn needs to be synchronized like this:

`private synchronized CharSequence getSubstitutionChars(int codePoint) {	`

bertfrees replied:
@carl-textalk It can go in, but only the `retrieveSubstituteFromUserTables` method really needed a "synchronized" keyword. (Possibly `addSubstitutionTable` too however you can assume that this will only be called during initialization of the UCharReplacer object.) Personally I would have just moved `mSubstitutionTable` and `mSubstitutionTablesIterator` inside `retrieveSubstituteFromUserTables` though.

carl-textalk replied:
Okey, i thought that it would cause overhead to load the substitution-table on every sentence?

bertfrees replied:
The table is loaded only once (when `addSubstitutionTable` is called).

The global variables just have no use. They are reassigned in every `retrieveSubstituteFromUserTables` call. There is no overhead brought by making global variables local.

carl-textalk replied:
Ok then i will do as you suggested then.

bertfrees replied:
Ok.

carl-textalk commented on modules/tts/tts-adapters/tts-adapter-cereproc/src/main/java/org/daisy/pipeline/tts/cereproc/impl/CereProcService.java:
```diff
@@ -103,7 +104,7 @@ protected CereProcEngine newEngine(
 				}
 			}
-			return new CereProcEngine(CereProcEngine.Variant.STANDARD, this, server, port, client, priority);
+			return new CereProcEngine(CereProcEngine.Variant.STANDARD, this, server, port, client, priority, this.getSSMLxslTransformerURL());
```
Hmm.. im not sure its just that simple to override the function, That causes a NullPointer when i tried it. 
I will have to look deeper in to what exactly causes the error.

bertfrees replied:
There shouldn't be a NullPointerException because I did this change: https://github.com/daisy/pipeline-modules/commit/4300e0c2cd0cb44d01630b88c70909625fd53f9c. If you get an error please look into it and find out what I did wrong.

carl-textalk replied:
Ah! i figured it out. it was my own mistake.

bertfrees commented on modules/tts/tts-adapters/tts-adapter-cereproc/src/main/java/org/daisy/pipeline/tts/cereproc/impl/util/UCharReplacer.java:
```diff
@@ -0,0 +1,206 @@
+	private ArrayList<Map<Integer,String>> mSubstitutionTables; 					//ArrayList<HashMap:<int codepoint>,<replaceString>>: all loaded translationtables
+	private Map<Integer,String> mSubstitutionTable = null;						//represents the currently selected translationTable
+	private Iterator<Map<Integer,String>> mSubstitutionTablesIterator = null;			//recycled tables map iterator
+	private Normalizer2 normalizer;
```
Not super critical, but I wonder whether we still need the Normalizer2 class. As far as I know you can do NFD normalization with plain Java now. Same for UCharacterIterator. Eliminating dependencies is always a good thing.

carl-textalk replied:
Hi, i cant find UCharacterIterator in plain java, can you please provide which class you where referring to?

carl-textalk replied:
Removed dependency to com.ibm.icu

bertfrees commented on modules/tts/tts-adapters/tts-adapter-cereproc/src/test/java/org/daisy/pipeline/tts/cereproc/impl/CereProcEngineTest.java:
```diff
@@ -0,0 +1,77 @@
+import java.util.*;
+
+
+public class CereProcEngineTest extends AbstractTest {
```
@carl-textalk Oh, one thing I wanted to say but forgot. I wonder whether the thing also works with OSGi. Would be good to know. Can you try it? Just tun the tests with the "test-with-paxexam" Maven profile. You probably need to override the `testDependencies` method in this class to something like:

```java
@Override
protected String[] testDependencies() {
	return new String[] {
		pipelineModule("tts-common"),
		pipelineModule("audio-common"),
		pipelineModule("audio-encoder-lame"),
		"com.ibu.uci:icu4j:?",
	};
}
```

carl-textalk replied:
My two tests does not extend AbstractTest anymore.

carl-textalk commented on modules/tts/tts-adapters/tts-adapter-cereproc/src/main/java/org/daisy/pipeline/tts/cereproc/impl/CereProcEngine.java:
```diff
@@ -202,4 +235,72 @@ public TTSResource allocateThreadResources() throws SynthesisException, Interrup
+
+			XdmNode ssmlOut = (XdmNode) ssmlProcessed.get(0);
+			ThreadUnsafeXslTransformer transformer = getTransformer(ssmlOut.getUnderlyingNode().getConfiguration());
+			mTransformParams.put("voice", v.name);
```
changed it to method variable.

bertfrees commented on modules/tts/tts-adapters/tts-adapter-cereproc/src/main/java/org/daisy/pipeline/tts/cereproc/impl/util/UCharReplacer.java:
```diff
@@ -0,0 +1,195 @@
+ * 
+ * <p>Note - there is a significant difference between a unicode codepoint (32 bit int)
+ * and a UTF16 codeunit (=char) - a codepoint consists of one or two codeunits.</p>  
+ * <p>To make sure an int represents a codepoint and not a codeunit.
```
This sentence got cut short?

carl-textalk replied:
oh sorry, thats a mistake. I was gonna remove parts mentioning com.ibm.icu.text-related stuff. I will fix that

bertfrees commented:
Moved to https://github.com/daisy/pipeline-modules/pull/40 and https://github.com/daisy/pipeline-assembly/pull/196
